graphistry/collapse.py

```diff
@@ -0,0 +1,532 @@
+    :param attribute: `attribute` to collapse in `column` over
+    :param column: `column` to collapse over
+    :param directed:
+    """
+    g2 = get_children(g, node_id, hops=1)
+    e = get_edges_of_node(
+        g, node_id, directed=directed
+    )  # False just includes the src node
+    ndf, edf, src, dst, node = unpack(g2)
+    tdf = ndf[ndf[column] == attribute]
+    if not tdf.empty:
+        # Get edges that are not in attribute (outside edges)
+        outcluster = set(e.values).difference(set(tdf.node.values))
+        # get edges that are internal to attribute, we will use these later to collapse those edges to supernode
+        incluster = set(tdf.node.values).intersection(set(e.values))
+        if VERBOSE:
+            if len(outcluster):
+                print(
+                    f"{outcluster} are edges *not* in [[ {column}:{attribute} ]] for node {node_id}"
+                )
+                # get directionality
+            if len(incluster):
+                print(
+                    f"{incluster} are edges in [[ {column}:{attribute} ]] for node {node_id}"
+                )
+        return outcluster, incluster, tdf
+    return None, None, None
+
+
+def get_cluster_store_keys(ndf, node):
+    """
+        Main innovation in finding and adding to super node.
+        Checks if node is a segment in any collapse_node in COLLAPSE column of nodes DataFrame
+
+    --------------------------------------------------------------------------------------------
+
+    :param ndf: node DataFrame
+    :param node: node to find
+    :returns DataFrame of bools of where `wrap_key(node)` exists in COLLAPSE column
+    """
+    node = wrap_key(node)
+    return ndf[COLLAPSE_NODE].astype(str).str.contains(node, na=False)
+
+
+def in_cluster_store_keys(ndf, node) -> bool:
+    """
+        checks if node is in collapse_node in COLLAPSE column of nodes DataFrame
+
+    ------------------------------------------------------------------------------
+
+    :param ndf: nodes DataFrame
+    :param node: node to find
+    :returns bool
+    """
+    return any(get_cluster_store_keys(ndf, node))
+
+
+def reduce_key(key) -> str:
+    """
+        Takes "1 1 2 1 2 3" -> "1 2 3
+
+    ---------------------------------------------------
+
+    :param key: node name
+    :returns new node name with duplicates removed
+    """
+    uniques = " ".join(np.unique(key.split()))
+    return uniques
+
+
+def unwrap_key(name) -> str:
+    """
+        Unwraps node name: ~name~ -> name
+
+    ----------------------------------------
+
+    :param name: node to unwrap
+    :returns unwrapped node name
+    """
+    return str(name).replace(WRAP, "")
+
+
+def wrap_key(name) -> str:
+    """
+        Wraps node name -> ~name~
+
+    -----------------------------------
+
+    :param name: node name
+    :returns wrapped node name
+    """
+    name = str(name)
+    if WRAP in name:  # idempotency
+        return name
+    return f"{WRAP}{name}{WRAP}"
+
```

Comment:
so if `"~"` is in a node id, what happens?

Reply:
The age old problem of escaping/split/replace, open to suggestions -- might even have WRAP be parameter

`g.collapse(..., escaping_token='~')` 

?

Reply:
(that's ok, but maybe we can detect-and-throw at the beginning if that's the case?)